docs/programmers_guide/witness_formal_spec.md

```diff
@@ -245,4 +386,109 @@ returns the first value in the specified array
 
 ### `REST(array)`
 
-returns the array w/o the first item
\ No newline at end of file
+returns the array w/o the first item
+
+### `KECCAK(bytes)`
+
+returns a keccak-256 hash of `bytes`
+
+
+## Serialization
+
+The format for serialization of everything except hashes (that we know the
+length of) is [CBOR](https://cbor.io). It is RFC-specified and concise.
+
+For hashes we use the optimization of knowing the lengths, so we just read 32
+bytes
+
+### Block Witness Format
+
+Each block witness consists of a header followed by a list of operators.
+
+There is no length of witness specified anywhere, the code expects to just reach `EOF`.
+
+Serialized Witness: `(HEADER, OP1, OP2, ..., OPn-1, OPn, EOF)`
+
+#### Encoding
+
+##### CBOR
+
+The parts of the key that are encoded with CBOR are marked by the `CBOR` function.
+
+##### Keys
+
+Keys are also using custom encryption to make them more compact.
+
+The nibbles of a key are encoded in a following way `[FLAGS NIBBLE1+NIBBLE2 NIBBLE3+NIBBLE4 NIBBLE5... ]`
+
+*FLAGS*
+* bit 0 -- 1 if the number of nibbles were odd
+* bit 1 -- 1 if the nibbles end with 0x10 (the terminator byte)
+
+This is shown later as `ENCODE_KEY` function.
+
+#### Header
+
+format: `version:byte`
+
+encoded as `[ version ]`
+
+the current version is 1.
+
+#### Instructions 
+
+Each instruction starts with an opcode (`uint`).
+
+Then it might contain some data.
+
+##### `HASH` 
+
+format: `HASH hash:[32]byte`
+
+encoded as `[ 0x03 hash_byte_1 ... hash_byte_32 ]`
+
+##### `BRANCH`
+
+format: `BRANCH mask:uint32`
+
+*mask* defines which children are present 
+(e.g. `0000000000001011` means that children 0, 1 and 3 are present and the other ones are not)
+
+encoded as `[ 0x02 CBOR(mask)...]`
+
+##### `CODE`
+
+format: `CODE code:[]byte`
+
+encoded as `[ 0x04 CBOR(code)... ]`
+
+##### `EXTENSION`
+
+format: `EXTENSION key:[]byte` 
+
+encoded as `[ 0x01 CBOR(ENCODE_KEY(key))... ]`
+
+##### `LEAF`
+
+format: `LEAF key:[]byte value:[]byte` 
+
+encoded as `[ 0x00 CBOR(ENCODE_KEY(key))... CBOR(value)... ]`
+
+##### `ACCOUNT_LEAF`
+
+format: `ACCOUNT_LEAF key:[]byte flags [nonce:uint64] [balance:[]byte]` 
+
+encoded as `[ 0x05 CBOR(ENCODE_KEY(key))... flags /CBOR(nonce).../ /CBOR(balance).../ ]`
```

Comment:
Some people may wonder why the opcode integers are not written in the natural order, but 3, 4, 1, 0, 5? :)